### DISCUSSION TITLE
Using hooks to add a protected file url to a view response

### DISCUSSION THREAD

**Discussion by Silfalion** [pocketbase/pocketbase#4003] [Q&A] [answered]
Hello! The more I'm reading about PocketBase the more I'm falling for it, thank you so much!

So I wanted to always return a protected file url for a view request a on view by using the OnRecordViewRequest hook. 

- check if the user is authenticated
- create a protected file url
- add the file url as a new field in the response

Is this feasible?

I looked around but haven't found an answer to this so I wanted to be sure. Thank you.

**Comment by ganigeorgiev** [comment]
I'm not sure that I understand what do you mean by "_protected file url_", but you can mark a `file` field as protected and it will require a short-lived file token to be accessed - https://pocketbase.io/docs/files-handling/#protected-files.

If you want to hide certain fields from a collection records response a better idea could be to move the protected/hidden fields to their own collection with its own API rules that could reference back to the original with a `relation`.

**Reply by Silfalion** [reply]
Ah sorry, I'll explain it further. So right now I'm storing the download url of an image in a collection so that the client can directly use it instead of having to fetch the url of the image.

However I'd like the url to have a short lived token, be signed, though that would require manual fetching on the client side.

So I was wondering if I could replace the download url in the collection by the id of the image, then on the OnRecordViewRequest hook, generate a signed url and inject it in the response in place of the media id that would've been returned otherwise.

So basically, is it possible to return a view of a record containing the(freshly generated) signed url of an image instead of returning the id of the media which the client would use to generate its signed url?

And thank you for pointing out the relation-solution, I hadn't realized those could be used.

**Reply by ganigeorgiev** [reply, marked as answer]
> download url of an image in a collection so that the client can directly use it instead of having to fetch the url of the image

If the "image" in the above you mean a PocketBase `file` field value, note that the SDK client doesn't send requests to fetch the url of the image, it just constructs it from the collecitonId, recordId and filename values.

> is it possible to return a view of a record containing the(freshly generated) signed url of an image instead of returning the id of the media which the client would use to generate its signed url?

You can but we don't have a "OnSerialize" hook at the moment and you'll have to add your new field in every action where you want to be accessible (_and eventually perform the necessary access checks_). For the view action, this could look something like (using the JS pb_hooks):

```js
// pb_hooks/main.pb.js

onRecordViewRequest((e) => {
    // https://pocketbase.io/docs/js-routing/#retrieving-the-current-auth-state
    const admin      = e.httpContext.get("admin")
    const authRecord = e.httpContext.get("authRecord")
    if (admin || !authRecord) {
        return // ignore for admins and non-authenticated users
    }

    let fileToken = $tokens.recordFileToken($app, authRecord)

    let url = [
        $app.settings().meta.appUrl,
        e.record.collection().id,
        e.record.id,
        e.record.get("yourFileField"),
    ].join("/") + `?token=${fileToken}`

    e.record.withUnknownData(true) // allow serialization of fields not part of the collection schema
    e.record.set("protectedFileUrl", url)
})
```

**Reply by Silfalion** [reply]
> If the "image" in the above you mean a PocketBase file field value, note that the SDK client doesn't send requests to fetch the url of the image, it just constructs it from the collecitonId, recordId and filename values.

If the file is protected, wouldn't it require the client to fetch a signed url to be able to access the file/image like in this code:

```dart

import 'package:pocketbase/pocketbase.dart';

final pb = PocketBase('http://127.0.0.1:8090');

...

// authenticate
await pb.collection('users').authWithPassword('test@example.com', '1234567890');

// generate a file token
final fileToken = await pb.files.getToken();

// retrieve an example protected file url (will be valid ~5min)
final record = await pb.collection('example').getOne('RECORD_ID');
final url = pb.files.getUrl(record, record.getStringValue('myPrivateFile'), token: fileToken);
```

> You can but we don't have a "OnSerialize" hook at the moment and you'll have to add your new field in every action where you want to be accessible (and eventually perform the necessary access checks). For the view action, this could look something like (using the JS pb_hooks):

```js
// pb_hooks/main.pb.js

onRecordViewRequest((e) => {
    // https://pocketbase.io/docs/js-routing/#retrieving-the-current-auth-state
    const admin      = e.httpContext.get("admin")
    const authRecord = e.httpContext.get("authRecord")
    if (admin || !authRecord) {
        return // ignore for admins and non-authenticated users
    }

    let fileToken = $tokens.recordFileToken($app, authRecord)

    let url = [
        $app.settings().meta.appUrl,
        e.record.collection().id,
        e.record.id,
        e.record.get("yourFileField"),
    ].join("/") + `?token=${fileToken}`

    e.record.withUnknownData(true) // allow serialization of fields not part of the collection schema
    e.record.set("protectedFileUrl", url)
})
```

This is exactly what I was looking for! Thank you so much for taking the time to write the code, I was gonna fiddle with it after your confirmation, so that saves me time, really appreciate it. And OnSerialize would indeed make it easier, in my case, I just need to make fetching the image to rely more on the backend, so the onRecordViewRequest does a job well

**Reply by ganigeorgiev** [reply]
> If the file is protected, wouldn't it require the client to fetch a signed url to be able to access the file/image like in this code:...

A request is send only for the file token, but that token can be used with multiple files until it expires (_by default I think is 120sec; you can change it form the Admin UI > Settings > Token options_).

**Reply by Silfalion** [reply]
> A request is send only for the file token, but that token can be used with multiple files until it expires (by default I think is 120sec; you can change it form the Admin UI > Settings > Token options).

Ohhhh I see, so I have to manage the token life-cycle in only one centralized place in the app for it to work everywhere. That's very convenient. I'll have a look at that too then. Thank you so much!!
